doc/copp/CoPP Config and Management.md

```diff
@@ -0,0 +1,336 @@
+
+### copporch
+```copporch``` shall only be a consumer of APP DB CoPP Table. It is not expected to handle feature logic and the current handling of features like sFlow, NAT shall be revisited and removed to be added as part of ```coppmgr```. However `copporch` must be able to handle any new trap_id getting added or removed from an existing CoPP table, and handle attribute value set for a trap group
+
+### swssconfig
+Handling of CoPP config json file shall be removed from ```dockers/docker-orchagent/swssconfig.sh```
+
+### 00-copp.config.json
+This file shall be modified to be compatible to Config DB schema, currently placed at ```swssconfig/sample/00-copp.config.json```. It shall be renamed to ```copp_cfg.j2``` and moved to ```files/image_config/copp/copp_cfg.j2```
+
+### copp_cfg.json
+A new file ```copp_cfg.json``` shall be introduced to include default CoPP tables from copp_cfg.j2 placed under ```/etc/sonic```. This file shall be read by ```coppmgr``` during initialization. 
+
+### Default CoPP Tables
+There are two proposals for loading the default CoPP Tables.
+
+1. Add default CoPP Tables to be part of init_cfg.json. As part of the post start action, the Config DB shall be loaded with default CoPP tables and/or any previously configured CoPP tables as part of the following handler in ```files/build_templates/docker_image_ctl.j2```
+
+```
+    function postStartAction()
+        ...
+        if [ -r /etc/sonic/config_db.json ]; then
+            sonic-cfggen -j /etc/sonic/config_db.json --write-to-db
+            if [ -r /etc/sonic/init_cfg.json ]; then
+                sonic-cfggen -j /etc/sonic/init_cfg.json -j /etc/sonic/config_db.json --write-to-db
+            else
+                sonic-cfggen -j /etc/sonic/config_db.json --write-to-db
+            fi
+        fi
+```
+However, this approach has the following limitations:
+
+In warmboot scenarios, as init_cfg.json is not read during warmboot and may result in ***new*** CoPP Tables in the "to" image not getting applied.
+In coldboot, if user saves the config and the new release happened to have new CoPP values, it will not be applied since previous default values are present in config_db
+
+2. ```coppmgr``` reading directly from ```copp_cfg.json``` and apply the default CoPP Tables. 
+
```

Comment:
I have implemented a 3rd option that addresses a limitation of the first option so that we can load the default COPP configs in the ConfigDB to keep inline with SONiC architecture and some requirements for config CLIs.  

To address the issue of an old COPP default config stored in config_db.json file (on config save) overwriting the new image COPP default config with new values, we are comparing the default COPP config file with the ConfigDB entries in sonic-cfggen and only writing back config modifications to the config_db.json file.   No COPP config will be stored in config_db.json if no modifications have been made.   Only "user modifications" will be stored back in the config_db.json.   This change in sonic-cfggen is generic enough for other use cases as well.  Let me know if this 3rd option is a possibility for this feature.

Reply:
The thought is, lets not have default in config_db at all. In case of warmboot, we will again have the issue if entries are in config_db and new image has a different default value. In which case you've to add the similar logic in the warmboot script before saving the redis_db to skip saving default values. Also I think, sonic-cfggen should not be used to decide which TABLES must be written to config_db.json file vs not written to file.

Reply:
Ok, we can have a separate discussion on this specific issue.  The rest of the doc looks good.